presto-testing/src/main/java/io/prestosql/testing/datatype/CreateAndInsertDataSetup.java

```diff
@@ -71,16 +71,26 @@ private void insertRows(TestTable testTable, List<ColumnSetup> inputs)
 
     private TestTable createTestTable(List<ColumnSetup> inputs)
     {
-        return new TestTable(sqlExecutor, tableNamePrefix, "(" + columnDefinitions(inputs) + ")");
+        return new TestTable(sqlExecutor, tableNamePrefix, tableDefinition(inputs));
     }
 
-    private String columnDefinitions(List<ColumnSetup> inputs)
+    private String tableDefinition(List<ColumnSetup> inputs)
     {
-        List<String> columnTypeDefinitions = inputs.stream()
-                .map(input -> input.getDeclaredType().orElseThrow(() -> new IllegalArgumentException("declared type not set")))
-                .collect(toList());
-        Stream<String> columnDefinitions = range(0, columnTypeDefinitions.size())
-                .mapToObj(i -> format("col_%d %s", i, columnTypeDefinitions.get(i)));
-        return Joiner.on(",\n").join(columnDefinitions.iterator());
+        if (inputs.stream().allMatch(input -> input.getDeclaredType().isPresent())) {
+            return IntStream.range(0, inputs.size())
+                    .mapToObj(column -> format("col_%d %s", column, inputs.get(column).getDeclaredType().orElseThrow()))
+                    .collect(joining(",\n", "(\n", ")"));
+        }
+
+        return IntStream.range(0, inputs.size())
+                .mapToObj(column -> {
+                    ColumnSetup input = inputs.get(column);
+                    if (input.getDeclaredType().isEmpty()) {
+                        return format("%s AS col_%d", input.getInputLiteral(), column);
+                    }
+
+                    return format("CAST(%s AS %s) AS col_%d", input.getInputLiteral(), input.getDeclaredType().get(), column);
+                })
+                .collect(joining(",\n", "AS\nSELECT\n", "\nWHERE 'with no' = 'data'"));
```

Comment:
Non-standard inequality for creating empty tables but I like how literal it is.

Reply:
Alas! I used `1=0` too many times. Worn off.
Also, using something more unique allows to trace the source of the condition easily.